source/common/http/async_client_impl.cc

```diff
@@ -261,18 +266,21 @@ void AsyncStreamImpl::sendData(Buffer::Instance& data, bool end_stream) {
     return;
   }
 
-  if (buffered_body_ != nullptr) {
-    // TODO(shikugawa): Currently, data is dropped when the retry buffer overflows and there is no
-    // ability implement any error handling. We need to implement buffer overflow handling in the
-    // future. Options include configuring the max buffer size, or for use cases like gRPC
-    // streaming, deleting old data in the retry buffer.
-    if (buffered_body_->length() + data.length() > kBufferLimitForRetry) {
-      ENVOY_LOG_EVERY_POW_2(
-          warn, "the buffer size limit (64KB) for async client retries has been exceeded.");
-    } else {
-      buffered_body_->add(data);
+  if (!new_async_client_retry_logic_) {
+    if (buffered_body_ != nullptr) {
+      // TODO(shikugawa): Currently, data is dropped when the retry buffer overflows and there is no
+      // ability implement any error handling. We need to implement buffer overflow handling in the
+      // future. Options include configuring the max buffer size, or for use cases like gRPC
+      // streaming, deleting old data in the retry buffer.
+      if (buffered_body_->length() + data.length() > kDefaultDecoderBufferLimit) {
+        ENVOY_LOG_EVERY_POW_2(
+            warn, "the buffer size limit (64KB) for async client retries has been exceeded.");
+      } else {
+        buffered_body_->add(data);
+      }
```

Comment:
In the new implementation, we will not buffer the body here directly, but will let the router filter to decide whether the retry is possible (based on the `decoderBufferLimit()`) and whether the data should be buffered (the router will call the `addDecodedData()`)
This logic is more similar with our normal traffic.